fgpyo/util/metric.py

```diff
@@ -334,3 +344,330 @@ def fast_concat(*inputs: Path, output: Path) -> None:
+        file_format = MetricFileFormat(delimiter=delimiter)
+
+        assert_is_metric(metric_class)
+        io.assert_path_is_writeable(filepath)
+        if append:
+            io.assert_path_is_readable(filepath)
+            assert_file_header_matches_metric(filepath, metric_class, file_format)
+
+        self._metric_class = metric_class
+        self._fieldnames = _validate_output_fieldnames(
+            metric_class=metric_class,
+            include_fields=include_fields,
+            exclude_fields=exclude_fields,
+        )
+        self._fout = io.to_writer(filepath, append=append)
+        self._writer = DictWriter(
+            f=self._fout,
+            fieldnames=self._fieldnames,
+            delimiter=delimiter,
+        )
+
+        # If we aren't appending to an existing file, write the header before any rows
+        if not append:
+            self._writer.writeheader()
+
+    def __enter__(self) -> "MetricWriter":
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Type[BaseException],
+        exc_value: BaseException,
+        traceback: TracebackType,
+    ) -> None:
+        self.close()
+
+    def close(self) -> None:
+        """Close the underlying file handle."""
+        self._fout.close()
+
+    def write(self, metric: Metric) -> None:
+        """
+        Write a single Metric instance to file.
+
+        The Metric is converted to a dictionary and then written using the underlying
+        `csv.DictWriter`. If the `MetricWriter` was created using the `include_fields` or
+        `exclude_fields` arguments, the attributes of the dataclass are subset and/or reordered
+        accordingly before writing.
+
+        Args:
+            metric: An instance of the specified Metric.
+        """
+        if not isinstance(metric, self._metric_class):
+            raise ValueError(f"Must provide instances of {self._metric_class.__name__}")
+
+        # Serialize the Metric to a dict for writing by the underlying `DictWriter`
+        row = asdict(metric)
+
+        # Filter and/or re-order output fields if necessary
+        row = {fieldname: row[fieldname] for fieldname in self._fieldnames}
+
+        self._writer.writerow(row)
+
+    def writeall(self, metrics: Iterable[Metric]) -> None:
+        """
+        Write multiple Metric instances to file.
+
+        Each Metric is converted to a dictionary and then written using the underlying
+        `csv.DictWriter`. If the `MetricWriter` was created using the `include_fields` or
+        `exclude_fields` arguments, the attributes of each Metric are subset and/or reordered
+        accordingly before writing.
+
+        Args:
+            metrics: A sequence of instances of the specified Metric.
+        """
+        for metric in metrics:
+            self.write(metric)
+
+
+def _get_header(
+    reader: io.ReadableFileHandle,
+    file_format: MetricFileFormat,
+) -> Optional[MetricFileHeader]:
```

Comment:
If/when we implement a `MetricReader` in a similar style, you might consider refactoring to re-use some of the reading logic. I can imagine a `MetricReader` would wrap a readable file handle in a peekable iterator so you could consume the header but still be able to use the iterator with the header included for actual reading (maybe with a `DictReader`?)

Reply:
You won't need a peekable iterator, the handle will be left open to the first row of data, and metrics can be consumed after calling `_read_header`

proof of concept: https://github.com/msto/dataclass_io/blob/c3d1d622788e58c555efb2e266aa3622a95a6786/dataclass_io/reader.py#L57-L62